File: core/memory.py
"""Memory and temporal processing: short-term buffer, long-term memory
formation, emotional tagging, decay, cycle detection, history compression."""

from __future__ import annotations
import time
import numpy as np
from collections import deque


class Memory:
    """A single long-term memory."""
    __slots__ = ('timestamp', 'frame', 'state', 'intensity', 'valence',
                 'emotion', 'reinforcement_count', 'last_accessed')

    def __init__(self, timestamp, frame, state, intensity, valence, emotion):
        self.timestamp = timestamp
        self.frame = frame
        self.state = np.array(state, dtype=np.float32)
        self.intensity = float(intensity)
        self.valence = float(valence)
        self.emotion = str(emotion)
        self.reinforcement_count = 0
        self.last_accessed = frame

    def similarity(self, state_vec):
        dot = np.dot(self.state, state_vec)
        norm_a = np.linalg.norm(self.state)
        norm_b = np.linalg.norm(state_vec)
        if norm_a < 1e-9 or norm_b < 1e-9:
            return 0.0
        return float(dot / (norm_a * norm_b))

    def effective_strength(self, current_frame, half_life=2000.0):
        age = max(current_frame - self.frame, 1)
        decay = 2.0 ** (-age / half_life)
        emotion_mult = 1.0 + 0.5 * abs(self.valence)
        reinforce_mult = 1.0 + 0.1 * self.reinforcement_count
        return self.intensity * decay * emotion_mult * reinforce_mult


class HistoryBucket:
    """Compressed summary of a range of frames."""
    __slots__ = ('frame_start', 'frame_end', 'count', 'mean_state',
                 'var_state', 'mean_valence', 'max_intensity')

    def __init__(self, frame_start, frame_end, count, mean_state,
                 var_state, mean_valence, max_intensity):
        self.frame_start = frame_start
        self.frame_end = frame_end
        self.count = count
        self.mean_state = mean_state
        self.var_state = var_state
        self.mean_valence = mean_valence
        self.max_intensity = max_intensity


class MemorySystem:
    """Memory and temporal processing layer.

    Call process_frame() each tick with current weights and responses.
    Returns a bias vector to add to weights.
    """

    def __init__(self, aspects: list[str], *, stm_size: int = 50,
                 ltm_capacity: int = 500, compression_interval: int = 500,
                 influence_scale: float = 0.02, decay_half_life: float = 2000.0):
        self.aspects = list(aspects)
        self.n = len(aspects)
        self._idx = {a: i for i, a in enumerate(aspects)}

        self.stm: deque = deque(maxlen=stm_size)
        self.memories: list[Memory] = []
        self.ltm_capacity = ltm_capacity
        self.influence_scale = influence_scale
        self.decay_half_life = decay_half_life

        self._raw_history: list = []
        self._compressed: list[HistoryBucket] = []
        self._last_compress_frame = 0
        self.compression_interval = compression_interval

        self._cycle_accum = np.zeros(self.n)
        self._cycle_counts = np.zeros(self.n, dtype=int)
        self._detected_periods: dict[str, float] = {}

        self.frame = 0
        self._persist_buffer: deque = deque(maxlen=20)
        self._last_memory_frame = -100

    def process_frame(self, weights, responses) -> np.ndarray:
        """Main entry point. Returns additive bias from memory influence."""
        state = np.asarray(weights, dtype=np.float32)
        resp = np.asarray(responses, dtype=np.float32)
        self.frame += 1

        self.stm.append(state.copy())

        intensity = float(np.linalg.norm(resp))
        valence = self._compute_valence(state)
        emotion = self._classify_emotion(state)

        self._raw_history.append((self.frame, state.copy(), valence))
        self._persist_buffer.append(valence)

        if self._should_form_memory(intensity, valence):
            self._form_memory(state, intensity, valence, emotion)

        self._reinforce_similar(state)
        self._update_cycle_detection(state)

        if self.frame - self._last_compress_frame >= self.compression_interval:
            self._compress_history()

        return self._compute_memory_bias(state)

    def compute_influences(self, weight_dict: dict[str, float]) -> dict[str, float]:
        """Protocol-compatible: return per-aspect influence dict."""
        state = np.array([weight_dict.get(a, 0.0) for a in self.aspects], dtype=np.float32)
        bias = self._compute_memory_bias(state)
        return {a: float(bias[i]) for i, a in enumerate(self.aspects)}

    def get_stm_trend(self, window=10):
        if len(self.stm) < max(window, 2):
            return np.zeros(self.n)
        recent = np.array(list(self.stm)[-window:])
        x = np.arange(window, dtype=float)
        x -= x.mean()
        denom = np.dot(x, x)
        if denom < 1e-12:
            return np.zeros(self.n)
        slopes = np.zeros(self.n)
        for j in range(self.n):
            slopes[j] = np.dot(x, recent[:, j] - recent[:, j].mean()) / denom
        return slopes

    def get_detected_cycles(self):
        return dict(self._detected_periods)

    def get_strongest_memories(self, k=5):
        scored = [(m, m.effective_strength(self.frame, self.decay_half_life))
                  for m in self.memories]
        scored.sort(key=lambda x: x[1], reverse=True)
        return scored[:k]

    # ── Internals ──

    def _compute_valence(self, state):
        pos = ['self-esteem', 'motivation', 'agency', 'self-efficacy']
        v = sum(state[self._idx[a]] for a in pos if a in self._idx)
        if 'self-regulation' in self._idx:
            v -= 0.3 * abs(state[self._idx['self-regulation']])
        return float(np.tanh(v / max(len(pos), 1)))

    def _classify_emotion(self, state):
        def _get(name):
            return state[self._idx[name]] if name in self._idx else 0

        ea, se, mo, ag = _get('emotional_awareness'), _get('self-esteem'), \
                          _get('motivation'), _get('agency')
        sr = _get('self-regulation')

        if ea > 0.3 and mo > 0.2 and se > 0.1:
            return 'excited'
        if se > 0.3 and ag > 0.2:
            return 'confident'
        if mo > 0.3 and ag > 0.1:
            return 'driven'
        if se < -0.2 and mo < 0:
            return 'distressed'
        if ea < -0.2 and sr > 0.3:
            return 'suppressed'
        if ea < -0.1 and se < -0.1:
            return 'anxious'
        if abs(ea) < 0.15 and abs(mo) < 0.15:
            return 'calm'
        return 'neutral'

    def _should_form_memory(self, intensity, valence):
        if self.frame - self._last_memory_frame < 30:
            return False
        if intensity > 1.5:
            return True
        if len(self._persist_buffer) >= 15:
            recent = list(self._persist_buffer)[-15:]
            mean_v = sum(recent) / len(recent)
            if abs(mean_v) > 0.25:
                consistent = sum(1 for v in recent if v * mean_v > 0)
                if consistent >= 12:
                    return True
        if intensity > 0.8 and abs(valence) > 0.4:
            return True
        return False

    def _form_memory(self, state, intensity, valence, emotion):
        mem = Memory(time.monotonic(), self.frame, state.copy(),
                     min(intensity, 3.0), valence, emotion)
        self.memories.append(mem)
        self._last_memory_frame = self.frame
        if len(self.memories) > self.ltm_capacity:
            self._evict_weakest()

    def _evict_weakest(self):
        if not self.memories:
            return
        worst_i = min(range(len(self.memories)),
                      key=lambda i: self.memories[i].effective_strength(
                          self.frame, self.decay_half_life))
        self.memories.pop(worst_i)

    def _reinforce_similar(self, state):
        for mem in self.memories:
            if mem.similarity(state) > 0.92:
                mem.reinforcement_count += 1
                mem.last_accessed = self.frame

    def _compute_memory_bias(self, current_state):
        bias = np.zeros(self.n, dtype=np.float64)
        for mem in self.memories:
            strength = mem.effective_strength(self.frame, self.decay_half_life)
            if strength < 1e-4:
                continue
            sim = mem.similarity(current_state)
            if abs(sim) < 0.1:
                continue
            direction = mem.state - current_state
            sign = 1.0 if mem.valence >= 0 else -1.0
            magnitude = self.influence_scale * strength * abs(sim) * sign
            bias += magnitude * direction

        norm = np.linalg.norm(bias)
        if norm > 0.05:
            bias *= 0.05 / norm
        return bias

    def _update_cycle_detection(self, state):
        if len(self.stm) < 10:
            return
        recent = np.array(list(self.stm)[-10:])
        means = recent.mean(axis=0)
        for j in range(self.n):
            centered = recent[:, j] - means[j]
            crossings = sum(1 for k in range(1, len(centered))
                            if centered[k-1] * centered[k] < 0)
            self._cycle_accum[j] += crossings
            self._cycle_counts[j] += 1

        if self.frame % 100 == 0:
            self._detected_periods.clear()
            for j in range(self.n):
                if self._cycle_counts[j] > 0:
                    avg = self._cycle_accum[j] / self._cycle_counts[j]
                    if avg > 0.5:
                        self._detected_periods[self.aspects[j]] = round(20.0 / avg, 1)
            self._cycle_accum[:] = 0
            self._cycle_counts[:] = 0

    def _compress_history(self):
        cutoff = self.frame - self.compression_interval
        to_compress = [e for e in self._raw_history if e[0] <= cutoff]
        self._raw_history = [e for e in self._raw_history if e[0] > cutoff]
        self._last_compress_frame = self.frame

        bucket_size = 100
        for start in range(0, len(to_compress), bucket_size):
            chunk = to_compress[start:start + bucket_size]
            states = np.array([c[1] for c in chunk], dtype=np.float64)
            valences = [c[2] for c in chunk]
            self._compressed.append(HistoryBucket(
                frame_start=chunk[0][0], frame_end=chunk[-1][0],
                count=len(chunk), mean_state=states.mean(axis=0),
                var_state=states.var(axis=0),
                mean_valence=sum(valences) / len(valences),
                max_intensity=float(np.max(np.linalg.norm(
                    states - states.mean(axis=0), axis=1)))))

File: native/src/memory.rs
/// Memory and temporal processing: short-term buffer, long-term memory
/// formation, emotional tagging, decay, cycle detection, and history compression.
///
/// Mirrors core/memory.py exactly.

use std::collections::{HashMap, VecDeque};

// ── Emotion encoding ─────────────────────────────────────────────────────────
// 0=neutral, 1=excited, 2=confident, 3=driven, 4=distressed, 5=suppressed,
// 6=anxious, 7=calm

pub struct Memory {
    pub frame: u64,
    pub state: Vec<f32>,
    pub intensity: f32,
    pub valence: f32,
    pub emotion: u8,
    pub reinforcement_count: u16,
    pub last_accessed: u64,
}

impl Memory {
    fn similarity(&self, state_vec: &[f32]) -> f64 {
        let dot: f64 = self.state.iter().zip(state_vec).map(|(&a, &b)| a as f64 * b as f64).sum();
        let norm_a: f64 = self.state.iter().map(|&x| (x as f64) * (x as f64)).sum::<f64>().sqrt();
        let norm_b: f64 = state_vec.iter().map(|&x| (x as f64) * (x as f64)).sum::<f64>().sqrt();
        if norm_a < 1e-9 || norm_b < 1e-9 {
            return 0.0;
        }
        dot / (norm_a * norm_b)
    }

    pub fn effective_strength(&self, current_frame: u64, half_life: f64) -> f64 {
        let age = (current_frame.saturating_sub(self.frame)).max(1) as f64;
        let decay = (2.0f64).powf(-age / half_life);
        let emotion_mult = 1.0 + 0.5 * (self.valence as f64).abs();
        let reinforce_mult = 1.0 + 0.1 * self.reinforcement_count as f64;
        self.intensity as f64 * decay * emotion_mult * reinforce_mult
    }
}

pub struct MemorySystem {
    aspects: Vec<String>,
    aspect_indices: HashMap<String, usize>,
    stm: VecDeque<Vec<f32>>,
    memories: Vec<Memory>,
    pub frame: u64,
    n: usize,
    ltm_capacity: usize,
    influence_scale: f64,
    decay_half_life: f64,
    persist_buffer: VecDeque<f64>,
    last_memory_frame: i64,
    cycle_accum: Vec<f64>,
    cycle_counts: Vec<u64>,
    detected_periods: Vec<(usize, f64)>,
    stm_size: usize,
}

impl MemorySystem {
    pub fn new(
        aspects: Vec<String>,
        stm_size: usize,
        ltm_capacity: usize,
        influence_scale: f64,
        decay_half_life: f64,
    ) -> Self {
        let n = aspects.len();
        let aspect_indices: HashMap<String, usize> =
            aspects.iter().enumerate().map(|(i, a)| (a.clone(), i)).collect();
        Self {
            aspects,
            aspect_indices,
            stm: VecDeque::with_capacity(stm_size + 1),
            memories: Vec::new(),
            frame: 0,
            n,
            ltm_capacity,
            influence_scale,
            decay_half_life,
            persist_buffer: VecDeque::with_capacity(21),
            last_memory_frame: -100,
            cycle_accum: vec![0.0; n],
            cycle_counts: vec![0u64; n],
            detected_periods: Vec::new(),
            stm_size,
        }
    }

    /// Main entry point. Returns additive memory bias (length n).
    pub fn process_frame(&mut self, weights: &[f64], stimuli: &[f64]) -> Vec<f64> {
        let state: Vec<f32> = weights.iter().map(|&x| x as f32).collect();
        self.frame += 1;

        // STM push (bounded)
        if self.stm.len() == self.stm_size {
            self.stm.pop_front();
        }
        self.stm.push_back(state.clone());

        let resp: Vec<f64> = stimuli.to_vec();
        let intensity = vec_norm_f64(&resp) as f32;
        let valence = self.compute_valence(&state);
        let emotion = self.classify_emotion(&state);

        // persist_buffer is maxlen=20
        if self.persist_buffer.len() == 20 {
            self.persist_buffer.pop_front();
        }
        self.persist_buffer.push_back(valence as f64);

        if self.should_form_memory(intensity as f64, valence as f64) {
            self.form_memory(state.clone(), intensity, valence, emotion);
        }

        self.reinforce_similar(&state);
        self.update_cycle_detection(&state);

        self.compute_memory_bias(&state)
    }

    /// Protocol-compatible: return per-aspect influence from a weight slice.
    pub fn compute_influences(&self, weights: &[f64]) -> Vec<f64> {
        let state: Vec<f32> = weights.iter().map(|&x| x as f32).collect();
        self.compute_memory_bias(&state)
    }

    // ── Internals ────────────────────────────────────────────────────────────

    fn get_idx(&self, name: &str) -> Option<usize> {
        self.aspect_indices.get(name).copied()
    }

    fn get_state_val(&self, state: &[f32], name: &str) -> f32 {
        self.get_idx(name).map(|i| state[i]).unwrap_or(0.0)
    }

    fn compute_valence(&self, state: &[f32]) -> f32 {
        let pos = ["self-esteem", "motivation", "agency", "self-efficacy"];
        let v: f32 = pos
            .iter()
            .filter_map(|&a| self.get_idx(a))
            .map(|i| state[i])
            .sum();
        let sr = self.get_state_val(state, "self-regulation");
        let adjusted = v - 0.3 * sr.abs();
        let divisor = pos.len().max(1) as f32;
        (adjusted / divisor).tanh()
    }

    fn classify_emotion(&self, state: &[f32]) -> u8 {
        let ea = self.get_state_val(state, "emotional_awareness");
        let se = self.get_state_val(state, "self-esteem");
        let mo = self.get_state_val(state, "motivation");
        let ag = self.get_state_val(state, "agency");

        if ea > 0.3 && mo > 0.2 && se > 0.1 {
            return 1; // excited
        }
        if se > 0.3 && ag > 0.2 {
            return 2; // confident
        }
        if mo > 0.3 && ag > 0.1 {
            return 3; // driven
        }
        if se < -0.2 && mo < 0.0 {
            return 4; // distressed
        }
        let sr = self.get_state_val(state, "self-regulation");
        if ea < -0.2 && sr > 0.3 {
            return 5; // suppressed
        }
        if ea < -0.1 && se < -0.1 {
            return 6; // anxious
        }
        if ea.abs() < 0.15 && mo.abs() < 0.15 {
            return 7; // calm
        }
        0 // neutral
    }

    fn should_form_memory(&self, intensity: f64, valence: f64) -> bool {
        if self.frame as i64 - self.last_memory_frame < 30 {
            return false;
        }
        if intensity > 1.5 {
            return true;
        }
        let buf_len = self.persist_buffer.len();
        if buf_len >= 15 {
            let recent: Vec<f64> = self.persist_buffer.iter().rev().take(15).copied().collect();
            let mean_v: f64 = recent.iter().sum::<f64>() / 15.0;
            if mean_v.abs() > 0.25 {
                let consistent = recent.iter().filter(|&&v| v * mean_v > 0.0).count();
                if consistent >= 12 {
                    return true;
                }
            }
        }
        if intensity > 0.8 && valence.abs() > 0.4 {
            return true;
        }
        false
    }

    fn form_memory(&mut self, state: Vec<f32>, intensity: f32, valence: f32, emotion: u8) {
        let mem = Memory {
            frame: self.frame,
            state,
            intensity: intensity.min(3.0),
            valence,
            emotion,
            reinforcement_count: 0,
            last_accessed: self.frame,
        };
        self.memories.push(mem);
        self.last_memory_frame = self.frame as i64;
        if self.memories.len() > self.ltm_capacity {
            self.evict_weakest();
        }
    }

    fn evict_weakest(&mut self) {
        if self.memories.is_empty() {
            return;
        }
        let frame = self.frame;
        let hl = self.decay_half_life;
        let worst = self
            .memories
            .iter()
            .enumerate()
            .min_by(|(_, a), (_, b)| {
                a.effective_strength(frame, hl)
                    .partial_cmp(&b.effective_strength(frame, hl))
                    .unwrap()
            })
            .map(|(i, _)| i)
            .unwrap();
        self.memories.swap_remove(worst);
    }

    fn reinforce_similar(&mut self, state: &[f32]) {
        let frame = self.frame;
        for mem in &mut self.memories {
            if mem.similarity(state) > 0.92 {
                mem.reinforcement_count = mem.reinforcement_count.saturating_add(1);
                mem.last_accessed = frame;
            }
        }
    }

    fn compute_memory_bias(&self, current_state: &[f32]) -> Vec<f64> {
        let mut bias = vec![0.0f64; self.n];
        for mem in &self.memories {
            let strength = mem.effective_strength(self.frame, self.decay_half_life);
            if strength < 1e-4 {
                continue;
            }
            let sim = mem.similarity(current_state);
            if sim.abs() < 0.1 {
                continue;
            }
            let sign = if mem.valence >= 0.0 { 1.0f64 } else { -1.0f64 };
            let magnitude = self.influence_scale * strength * sim.abs() * sign;
            for j in 0..self.n {
                let dir = mem.state[j] as f64 - current_state[j] as f64;
                bias[j] += magnitude * dir;
            }
        }

        // Clamp norm to 0.05
        let norm = vec_norm_f64(&bias);
        if norm > 0.05 {
            let scale = 0.05 / norm;
            for b in &mut bias {
                *b *= scale;
            }
        }
        bias
    }

    fn update_cycle_detection(&mut self, _state: &[f32]) {
        if self.stm.len() < 10 {
            return;
        }
        // Take the last 10 entries
        let start = self.stm.len() - 10;
        let recent: Vec<&Vec<f32>> = self.stm.iter().skip(start).collect();

        for j in 0..self.n {
            let mean: f32 = recent.iter().map(|s| s[j]).sum::<f32>() / 10.0;
            let centered: Vec<f32> = recent.iter().map(|s| s[j] - mean).collect();
            let crossings: usize = (1..centered.len())
                .filter(|&k| centered[k - 1] * centered[k] < 0.0)
                .count();
            self.cycle_accum[j] += crossings as f64;
            self.cycle_counts[j] += 1;
        }

        if self.frame % 100 == 0 {
            self.detected_periods.clear();
            for j in 0..self.n {
                if self.cycle_counts[j] > 0 {
                    let avg = self.cycle_accum[j] / self.cycle_counts[j] as f64;
                    if avg > 0.5 {
                        // round to 1 decimal place
                        let period = (20.0 / avg * 10.0).round() / 10.0;
                        self.detected_periods.push((j, period));
                    }
                }
            }
            for v in &mut self.cycle_accum {
                *v = 0.0;
            }
            for c in &mut self.cycle_counts {
                *c = 0;
            }
        }
    }

    // ── Accessors ─────────────────────────────────────────────────────────────

    pub fn detected_periods(&self) -> &[(usize, f64)] {
        &self.detected_periods
    }

    pub fn memory_count(&self) -> usize {
        self.memories.len()
    }

    pub fn aspect_name(&self, idx: usize) -> Option<&str> {
        self.aspects.get(idx).map(|s| s.as_str())
    }
}

fn vec_norm_f64(v: &[f64]) -> f64 {
    v.iter().map(|&x| x * x).sum::<f64>().sqrt()
}
